File: bindings/c/examples/screen_record.c
/*
 * screen_record.c — Mediaway screen capture C ABI reachability check.
 *
 * NOT a working screen recorder, and cannot become one from C in this pass: per
 * `crates/mediaway-device-ffi/adr/0001-capture-c-abi.md` § Finding 2 and §
 * Deferred, `WindowsScreenCapture` / `WindowsWindowCapture` are real,
 * hardware-verified, Zero-Copy Rust backends, but BOTH require a live
 * `GpuDeviceHandle::DirectX11(ID3D11Device*)` and reject `CpuFramesOk` outright —
 * there is no CPU fallback, and no C ABI representation for a GPU device handle
 * exists yet (the same open problem `mediaway-pipeline-ffi/adr/0001` already
 * deferred for `AutoVideoEncodeConfig::gpu_device`; see
 * `docs/spec/gpu-interop.md`). `mediaway_video_capture_open()` on a Screen-kind
 * config therefore ALWAYS returns `MEDIAWAY_DEVICE_STATUS_UNSUPPORTED` today —
 * real, deterministic, documented behavior (`include/mediaway/device.h`'s own
 * file header), not a bug this example works around.
 *
 * This file previously sketched a full screen+mic capture -> encode -> MP4 flow
 * (the same shape as the corrected `bindings/c/examples/camera_record.c`) before
 * the real ABI existed. That flow can never succeed from C until a follow-up ADR
 * designs a GPU-device-handle C representation, so it has been replaced with
 * exactly what the real ABI does today: open, observe UNSUPPORTED, and exit
 * gracefully with an explanation. This is the one part of this example that is
 * real, buildable, and link+run-verified in this dev environment (see
 * `docs/ai/wiki/device/ffi-c-abi.md`'s "Building the C examples on Windows"
 * section) — it demonstrates a real capability gap, not a working capture path.
 */

#include <mediaway/device.h>

#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

int main(void) {
    const uint32_t fps = 30;
    const mediaway_rational_t tb = {.num = 1, .den = fps};

    /* Display index 0 = primary display. mediaway_video_capture_open() on a
     * Screen-kind config ALWAYS returns MEDIAWAY_DEVICE_STATUS_UNSUPPORTED in
     * this pass — see the file header. */
    mediaway_video_capture_config_t video_cfg = mediaway_video_capture_config_screen(0, tb);
    mediaway_video_capture_t *video = NULL;
    mediaway_device_status_t status = mediaway_video_capture_open(&video_cfg, &video);

    if (status != MEDIAWAY_DEVICE_STATUS_UNSUPPORTED) {
        /* A surprise: this file's whole premise (see the header comment) is that
         * screen capture cannot be opened from C in this pass. Any other status —
         * including MEDIAWAY_DEVICE_STATUS_OK — means the real ABI's behavior
         * changed and this example is out of date. */
        fprintf(stderr,
                "screen_record: unexpected status %d opening screen capture (expected "
                "MEDIAWAY_DEVICE_STATUS_UNSUPPORTED) — this example is out of date, see "
                "the file header\n",
                (int)status);
        if (video != NULL) {
            mediaway_video_capture_close(video);
        }
        return EXIT_FAILURE;
    }

    printf("screen_record: screen capture is UNSUPPORTED from C in this pass (status %d) — "
           "WindowsScreenCapture (and WindowsWindowCapture) require a live GPU device handle "
           "(GpuDeviceHandle::DirectX11) with no CPU fallback, and there is no C ABI "
           "representation for a GPU device handle yet. See "
           "crates/mediaway-device-ffi/adr/0001-capture-c-abi.md Finding 2 and Deferred. "
           "Nothing to record — exiting.\n",
           (int)status);
    return EXIT_SUCCESS;
}

File: bindings/nodejs/examples/screen-record.ts
/**
 * Screen + mic capture -> encode -> fragmented MP4 — aspirational quick-start example.
 *
 * ASPIRATIONAL EXAMPLE: no `@mediaway/device` or `@mediaway/encoder` npm package
 * exists yet. This file shows the target ergonomics for a future Node.js binding
 * over Mediaway's C ABI (napi-rs native addon under the hood, wrapped in
 * idiomatic Node/TypeScript: typed configs, `Error` subclasses, explicit
 * `close()`). This is the Node.js host reaching native code through the C ABI —
 * Node runs on servers/desktops and calls OS screen/mic capture APIs through the
 * native addon; it does not use browser capture APIs. See ../README.md and
 * docs/spec/c-ffi.md.
 *
 * Builds a screen-recording pipeline from the same building blocks as the
 * auto-encode example (config -> open auto encoder -> open encode session ->
 * writeFrame -> finish), plus a device-capture layer, glued together by one
 * small, platform-agnostic `record()` function that only knows about the
 * `VideoCapture`/`AudioCapture` interfaces — never the concrete OS backend.
 *
 * Run (once the real packages exist):
 *   npx tsx screen-record.ts
 */

import { writeFile } from "node:fs/promises";
import type { Rational } from "@mediaway/common";
import {
  type AudioCapture,
  type VideoCapture,
  openMicrophoneCapture,
  openScreenCapture,
} from "@mediaway/device";
import {
  AutoVideoEncodeConfig,
  EncodeSession,
  EncoderUnavailableError,
  openAutoEncoder,
  type VideoFrame,
} from "@mediaway/encoder";

const FPS = 30;
const DURATION_MS = 3_000;
const OUTPUT_PATH = "out_screen.mp4";

/** Build one grey NV12 frame: width*height Y bytes, then width*height/2 interleaved UV bytes. */
function makeGreyNv12Frame(width: number, height: number, pts: number): VideoFrame {
  const ySize = width * height;
  const uvSize = ySize / 2;
  const data = Buffer.alloc(ySize + uvSize, 0x80); // Y=128, U=128, V=128 everywhere

  return {
    pts,
    duration: 1,
    width,
    height,
    pixelFormat: "nv12",
    data,
  };
}

/**
 * Poll video + (optional) audio capture and feed the encode session until
 * `durationMs` elapses.
 *
 * This function is the reusable "record loop": it only depends on the
 * `VideoCapture`/`AudioCapture` interfaces and an already-open `EncodeSession`,
 * so it works unchanged regardless of which OS backend opened the captures.
 */
async function record(
  video: VideoCapture,
  audio: AudioCapture | undefined,
  session: EncodeSession,
  durationMs: number,
): Promise<void> {
  const deadline = Date.now() + durationMs;
  let frameIndex = 0;

  while (Date.now() < deadline) {
    // A real capture frame may reference GPU-resident memory (Zero-Copy handle);
    // this example only writes a synthetic placeholder into the encoder instead
    // of the real pixels, but still models the release step below.
    const videoFrame = video.pollFrame();
    if (videoFrame !== undefined) {
      try {
        await session.writeFrame(makeGreyNv12Frame(video.width, video.height, frameIndex));
        frameIndex++;
      } finally {
        // Release the GPU-resident buffer back to the OS before polling again.
        videoFrame.release();
      }
    }

    // Drain any polled audio frames — this example doesn't wire audio into the
    // MP4 yet, just keeps the capture queue from backing up.
    if (audio !== undefined) {
      while (audio.pollFrame() !== undefined) {
        // intentionally discarded
      }
    }

    if (videoFrame === undefined) {
      // Nothing new yet — yield briefly instead of busy-spinning.
      await new Promise<void>((resolve) => setTimeout(resolve, 1));
    }
  }
}

async function main(): Promise<void> {
  let video: VideoCapture;
  try {
    // 0 = primary display; capture settles on its own actual stream geometry.
    video = await openScreenCapture({ displayIndex: 0, timeBase: { num: 1, den: FPS } satisfies Rational });
  } catch (err) {
    console.error(`screen-record: screen capture unavailable on this platform (${(err as Error).message})`);
    return;
  }

  let audio: AudioCapture | undefined;
  try {
    audio = await openMicrophoneCapture({ timeBase: { num: 1, den: 48_000 } satisfies Rational });
  } catch (err) {
    console.warn(`screen-record: microphone unavailable, continuing without audio (${(err as Error).message})`);
  }

  console.log(`screen-record: capturing at ${video.width}x${video.height}`);

  // Defaults for H.264 at the capture's real geometry @30fps, then override bitrate.
  const config: AutoVideoEncodeConfig = {
    ...AutoVideoEncodeConfig.defaults("h264", video.width, video.height, { num: 1, den: FPS }),
    bitrateBps: 8_000_000,
  };

  let encoder;
  try {
    // Tries the best available backend on this machine (GPU Zero-Copy first,
    // CPU-upload fallback) — throws if none exists here.
    encoder = await openAutoEncoder(config);
  } catch (err) {
    if (err instanceof EncoderUnavailableError) {
      console.error(`screen-record: no encoder backend available on this platform (${err.message})`);
      video.close();
      audio?.close();
      return;
    }
    throw err;
  }

  const session = new EncodeSession(encoder);
  try {
    try {
      await record(video, audio, session, DURATION_MS);
    } finally {
      video.close();
      audio?.close();
    }

    // Flushes the encoder, finalizes the muxer, and returns the complete MP4 bytes.
    const mp4Bytes = await session.finish();

    await writeFile(OUTPUT_PATH, mp4Bytes);

    console.log(
      `screen-record: ${(DURATION_MS / 1000).toFixed(0)}s @ ${video.width}x${video.height} -> ${OUTPUT_PATH} (${mp4Bytes.length} bytes)`,
    );
  } finally {
    session.close();
  }
}

await main();
